trumania/core/util_functions.py

```diff
@@ -89,23 +89,33 @@ def merge_2_dicts(dict1, dict2, value_merge_func=None):
     if dict1 is None:
         return dict2
 
-    def merged_value(key):
-        if key not in dict1:
-            return dict2[key]
-        elif key not in dict2:
-            return dict1[key]
-        else:
-            if value_merge_func is None:
-                raise ValueError(
-                    "Conflict in merged dictionaries: merge function not "
-                    "provided but key {} exists in both dictionaries".format(
-                        key))
-
-            return value_merge_func(dict1[key], dict2[key])
-
-    keys = set(dict1.keys()) | set(dict2.keys())
-
-    return {key: merged_value(key) for key in keys}
+    if dict1 == dict2:
+        for k, v in dict1.items():
+            dict1[k] = value_merge_func(v, v)
+        return dict1
+
+    dict1_set = set(dict1)
+    dict2_set = set(dict2)
+
+    keys_to_merge = dict1_set.intersection(dict2_set)
+
+    if len(keys_to_merge) == 0 and value_merge_func is None:
+        raise ValueError(
+            "Conflict in merged dictionaries: merge function not "
+            "provided but keys {} exists in both dictionaries".format(
+                keys_to_merge))
+
+    values_merged = dict()
+
+    for key_to_merge in keys_to_merge:
+        old_value1 = dict1.pop(key_to_merge)
+        old_value2 = dict2.pop(key_to_merge)
+
```

Comment:
Unless really proven to be necessary for optimisations reasons, I would discourage mutating `dict1` and `dict2` since this is a parameter we receive as argument. 
This is because, from the point of view of the caller, the statement `a = merge_2_dicts(b, c)` is not expected to update `b` nor `c`, otherwise the method becomes hard to reason about and hard to use it in more complex compositions. 

I think that thanks to the order you put in `return {**dict1, **dict2, **values_merged}` (nice one, I didn't know that syntax :) ), the `pop` is actually not necessary and gives priority to the values on the right => if we rephrase that for-loop as follows: 


```
values_merged = {k: value_merge_func(dict1[k], dict2[k] for k in key_to_merge}
```

We obtain the equivalent logic. 

What do you think ?